Implementation of a configurable github crawler.

damienr74 opened the pull request:
It is able to retreive all the kustomization files, from the github api
and should be able to incrementally find the newly indexed files if ever
there are too many kustomization files to retrieve based off of API
shortcommings.

To make this easier to read, use, and modify, I've abstracted the
important parts of the github query api into crawler/github/query.go
which allows to describe at a high level what is to be searched without
knowing the API syntax.

This is a pretty big commit, If you don't want to review it in a working
state, I really don't mind breaking it up into smaller commits. This is
just huge since we didn't know where to put the code, and I had been
working on this for a while..

damienr74 commented:
This is what the crawler looks like in a working state. As stated before, I really don't mind breaking the crawler file into multiple commits for code review purposes.

Liujingfang1 commented on internal/search/crawler/github/crawler.go:
```diff
@@ -0,0 +1,577 @@
+	log.LstdFlags|log.LUTC|log.Llongfile)
+
+// Implements crawler.Crawler.
+type githubCrawler struct {
```
Can you move split this file into another PR? I'll review the query parts in this PR. Thanks.

damienr74 replied:
Sure!

Liujingfang1 commented on internal/search/crawler/github/crawler.go:
```diff
@@ -0,0 +1,577 @@
+	"strings"
+	"time"
+
+	"github.com/damienr74/kustomize/internal/search/doc"
```
Can you change it to `sigs.k8s.io/kustomize`? Then when we move it to the kustomize repo later, we don't need to update the import path.

damienr74 replied:
The reason why it was set up this way was to be able to actually compile/test all of this code locally. With the wrong local path, I had issues with gomodules, but I can try again and see if I manage to get it working.

damienr74 replied:
This was a problem from before I figured out how to implement multiple modules in a repo. Fixed, thanks for bringing it up!

damienr74 committed:
Implementation of github query helper library.

To make this easier to read, use, and modify, I've abstracted the
important parts of the github query api into crawler/github/query.go
which allows to describe at a high level what is to be searched without
knowing the API syntax.

Liujingfang1 commented on internal/search/crawler/github/queries.go:
```diff
@@ -0,0 +1,229 @@
+}
+
+// Implementation detail, not important to external API.
+type queryField struct {
```
Move `queryField` definition to the top of this file.

damienr74 replied:
done.

Liujingfang1 commented on internal/search/crawler/github/queries.go:
```diff
@@ -0,0 +1,229 @@
+		value = fmt.Sprint(v)
+	}
+
+	if qf.name == "" {
```
Is this a valid case?

damienr74 replied:
Keywords don't have a name, they are formatted only using the value (the `{name}:` being implicit).

Liujingfang1 commented on internal/search/crawler/github/queries.go:
```diff
@@ -0,0 +1,229 @@
+func (q Query) String() string {
+	strs := make([]string, 0, len(q))
+	for _, elem := range q {
+		str := fmt.Sprint(elem)
```
Maybe change it to 
```
str := elem.String()
```

damienr74 replied:
sure

Liujingfang1 commented on internal/search/crawler/github/queries.go:
```diff
@@ -0,0 +1,229 @@
+
+// Filesize takes a rangeFormatter and formats it according to the Github API.
+func Filesize(r rangeFormatter) queryField {
+	return queryField{"size", r}
```
Please assign the fields explicitly.
```
return queryField{name: "size", value: r}
```

Liujingfang1 commented on internal/search/crawler/github/queries.go:
```diff
@@ -0,0 +1,229 @@
+
+// Filename takes a filename and formats it according to the Github API.
+func Filename(f string) queryField {
+	return queryField{"filename", f}
```
ditto

Liujingfang1 commented on internal/search/crawler/github/queries.go:
```diff
@@ -0,0 +1,229 @@
+
+// Path takes a filepath and formats it according to the Github API.
+func Path(p string) queryField {
+	return queryField{"path", p}
```
ditto

Liujingfang1 commented on internal/search/crawler/github/queries.go:
```diff
@@ -0,0 +1,229 @@
+// (patial) query, returns a request object with the (parital) query. Must call
+// the URL method to get the string value of the URL. See request.CopyWith, to
+// understand why the request object is useful.
+func (rc RequestConfig) CodeSearchRequestWith(queryParams ...queryField) request {
```
This is an exported function, but it accepts a private type. Can you change it to accept `Query` as the parameters.
```
CodeSearchRequestWith(query Query)
```

damienr74 replied:
Sure.

Liujingfang1 commented:
/lgtm
Feel free to merge it
